src/signature/commitment_signature.rs

```diff
@@ -0,0 +1,292 @@
+    /// completely random and never reused - that responsibility lies with the calling function.
+    ///   C = a*H + x*G          ... (Pedersen commitment to the value 'a' using blinding factor 'x')
+    ///   R = k_2*H + k_1*G      ... (a public (Pedersen) commitment nonce created with the two random nonces)
+    ///   u = k_1 + e.x          ... (the first publicly known private key of the signature signing with 'x')
+    ///   v = k_2 + e.a          ... (the second publicly known private key of the signature signing with 'a')
+    ///   signature = (R, u, v)  ... (the final signature tuple)
+    pub fn sign(
+        secret_a: K,
+        secret_x: K,
+        nonce_a: K,
+        nonce_x: K,
+        challenge: &[u8],
+    ) -> Result<Self, CommitmentSignatureError>
+    where
+        K: Add<Output = K> + Mul<P, Output = P> + Mul<Output = K>,
+    {
+        let e = match K::from_bytes(challenge) {
+            Ok(e) => e,
+            Err(_) => return Err(CommitmentSignatureError::InvalidChallenge),
+        };
+        let ea = e.clone() * secret_a;
+        let ex = e * secret_x;
+
+        let v = nonce_a.clone() + ea;
+        let u = nonce_x.clone() + ex;
+
+        let factory = C::default();
+        let public_commitment_nonce = factory.commit(&nonce_x, &nonce_a);
+
+        Ok(Self::new(public_commitment_nonce, u, v))
+    }
+
+    /// Verify if the commitment signature signed the commitment using the specified challenge (as bytes). If the
+    /// provided challenge n bytes cannot be converted to a secret key, this function also returns false.
+    pub fn verify_challenge<'a>(&self, public_commitment: &'a HomomorphicCommitment<P>, challenge: &[u8]) -> bool
+    where
+        for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+    {
+        let e = match K::from_bytes(&challenge) {
+            Ok(e) => e,
+            Err(_) => return false,
+        };
+
+        self.verify(public_commitment, &e)
+    }
+
+    /// Verify if the commitment signature signed the commitment using the specified challenge (as secret key).
+    ///   v*H + u*G = R + e.C
+    pub fn verify<'a>(&self, public_commitment: &'a HomomorphicCommitment<P>, challenge: &K) -> bool
+    where
+        for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+    {
+        let lhs = self.calc_signature_verifier();
+        let rhs = &self.public_commitment_nonce + &(public_commitment * challenge);
+        // Implementors should make this a constant time comparison
+        lhs == rhs
+    }
+
+    /// This function returns the complete signature tuple (R, u, v)
+    #[inline]
+    pub fn get_complete_signature_tuple(&self) -> (&HomomorphicCommitment<P>, &K, &K) {
```

Comment:
What is the value of this method?

Reply:
helper method so you dont have to call:
```
sig.u()
sig.v()
sig.nonce()
```

you can just call
`(_,_,_) = sig.get_complete_tuple`

Reply:
Note that by convention, Rust getters don't start with `get_`